zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs

```diff
@@ -118,89 +114,75 @@ impl InputSource for TransactionRecordsById {
-                );
+            match unselected.get(unselected_iterator) {
+                None => {
+                    // the iterator went off the end of the vector without finding a note big enough to complete the transaction... add the biggest note and reset the iteraton
+                    selected.push(unselected.pop().expect("nonempty"));
+                    unselected_iterator = 0;
+                    continue;
+                }
+                Some(smallest_unselected) => {
+                    // selected a note to test if it has enough value to complete the transaction on its own
+                    if smallest_unselected.1
+                        >= target_value.into_u64()
+                            - selected.iter().fold(0, |sum, (_id, value)| sum + *value)
+                    {
+                        selected.push(*smallest_unselected);
+                        unselected.remove(unselected_iterator);
+                        break;
+                    } else {
+                        // this note is not big enough. try the next
+                        unselected_iterator += 1;
+                    }
+                }
             }
         }
 
-        sapling_note_noteref_pairs.sort_by_key(|sapling_note| sapling_note.0.value().inner());
-        orchard_note_noteref_pairs.sort_by_key(|orchard_note| orchard_note.0.value().inner());
-
-        let mut sapling_notes = Vec::<ReceivedNote<NoteId, sapling_crypto::Note>>::new();
-        let mut orchard_notes = Vec::<ReceivedNote<NoteId, orchard::Note>>::new();
-        if let Some(missing_value_after_sapling) = sapling_note_noteref_pairs.into_iter().try_fold(
-            Some(target_value),
-            |rolling_target, (note, note_id)| match rolling_target {
-                Some(targ) if targ == NonNegativeAmount::ZERO => Ok(None),
-                Some(targ) => {
-                    sapling_notes.push(
-                        self.get(note_id.txid())
-                            .and_then(|tr| {
-                                tr.get_received_note::<SaplingDomain>(note_id.output_index() as u32)
-                            })
-                            .ok_or(InputSourceError::NoteCannotBeIdentified(note_id))?,
-                    );
-                    Ok(targ
-                        - NonNegativeAmount::from_u64(note.value().inner()).map_err(|e| {
-                            InputSourceError::OutputTooBig((note.value().inner(), e))
-                        })?)
+        let mut selected_sapling = Vec::<ReceivedNote<NoteId, sapling_crypto::Note>>::new();
+        let mut selected_orchard = Vec::<ReceivedNote<NoteId, orchard::Note>>::new();
+        selected
+            .iter()
+            .try_for_each(|(id, _value)| match id.protocol() {
+                zcash_client_backend::ShieldedProtocol::Sapling => {
+                    match self.get(id.txid()).and_then(|transaction_record| {
```

Comment:
Helperize?

Reply:
oh, i can use the transitive property to simplify this